server/src/main/java/org/elasticsearch/action/search/TransportSearchShardsAction.java

```diff
@@ -0,0 +1,139 @@
+    private final SearchService searchService;
+    private final RemoteClusterService remoteClusterService;
+    private final ClusterService clusterService;
+    private final SearchTransportService searchTransportService;
+    private final IndexNameExpressionResolver indexNameExpressionResolver;
+
+    @Inject
+    public TransportSearchShardsAction(
+        TransportService transportService,
+        SearchService searchService,
+        ActionFilters actionFilters,
+        ClusterService clusterService,
+        TransportSearchAction transportSearchAction,
+        SearchTransportService searchTransportService,
+        IndexNameExpressionResolver indexNameExpressionResolver
+    ) {
+        super(SearchShardsAction.NAME, transportService, actionFilters, SearchShardsRequest::new);
+        this.transportSearchAction = transportSearchAction;
+        this.searchService = searchService;
+        this.remoteClusterService = transportService.getRemoteClusterService();
+        this.clusterService = clusterService;
+        this.searchTransportService = searchTransportService;
+        this.indexNameExpressionResolver = indexNameExpressionResolver;
+    }
+
+    @Override
+    protected void doExecute(Task task, SearchShardsRequest searchShardsRequest, ActionListener<SearchShardsResponse> listener) {
+        final long relativeStartNanos = System.nanoTime();
+        SearchRequest original = searchShardsRequest.getSearchRequest();
+        final TransportSearchAction.SearchTimeProvider timeProvider = new TransportSearchAction.SearchTimeProvider(
+            original.getOrCreateAbsoluteStartMillis(),
+            relativeStartNanos,
+            System::nanoTime
+        );
+        ClusterState clusterState = clusterService.state();
+        Rewriteable.rewriteAndFetch(
+            original,
+            searchService.getRewriteContext(timeProvider::absoluteStartMillis),
+            ActionListener.wrap(searchRequest -> {
+                // No user preference defined in search request - apply cluster service default
+                if (searchRequest.allowPartialSearchResults() == null) {
+                    searchRequest.allowPartialSearchResults(searchService.defaultAllowPartialSearchResults());
+                }
+                Map<String, OriginalIndices> groupedIndices = remoteClusterService.groupIndices(
+                    searchRequest.indicesOptions(),
+                    searchRequest.indices()
+                );
+                OriginalIndices originalIndices = groupedIndices.remove(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY);
+                if (groupedIndices.isEmpty() == false) {
+                    throw new UnsupportedOperationException("search_shards API doesn't support remote indices " + searchRequest);
+                }
+                Index[] concreteIndices = transportSearchAction.resolveLocalIndices(originalIndices, clusterState, timeProvider);
```

Comment:
while it's good that we don't depend on transport search action to run the can match phase, it feels like we may be able to consolidate the common parts outside of transport search action, because we still depend on it here and below? Or do you think that's overkill given that only these two APIs will ever share this code?

Reply:
+1, We should refactor/cleanup the TransportSearchAction. However, I prefer to do it in a follow-up after we support PIT in this new API because we are going to make more changes in TransportSearchAction with PIT. Are you okay with that?

Reply:
sounds good to me